etk-asm/src/ingest.rs

```diff
@@ -403,70 +283,87 @@ where
+        for node in nodes {
             match node {
                 Node::Op(op) => {
-                    self.sources.write(RawOp::Op(op))?;
+                    raws.push(RawOp::Op(op));
                 }
-                Node::Raw(raw) => {
-                    self.sources.write(RawOp::Raw(raw))?;
+                Node::Import(imp_path) => {
+                    let new_raws = self.resolve_and_ingest(program, imp_path)?;
+                    raws.extend(new_raws);
                 }
-                Node::Import(path) => {
-                    let partial = self.sources.resolve(path, Scope::same())?;
-                    let parsed = parse_file(partial.path())?;
-                    partial.push(parsed);
+                Node::Include(inc_path) => {
+                    let inc_raws = self.resolve_and_ingest(program, inc_path)?;
+                    raws.push(RawOp::Scope(inc_raws));
                 }
-                Node::Include(path) => {
-                    let partial = self.sources.resolve(path, Scope::independent())?;
-                    let parsed = parse_file(partial.path())?;
-                    partial.push(parsed);
-                }
-                Node::IncludeHex(path) => {
-                    let partial = self.sources.resolve(path, Scope::same())?;
-
+                Node::IncludeHex(hex_path) => {
                     let file =
-                        std::fs::read_to_string(partial.path()).with_context(|_| error::Io {
-                            message: "reading hex include",
-                            path: partial.path().to_owned(),
+                        std::fs::read_to_string::<&PathBuf>(&hex_path).with_context(|_| {
+                            error::Io {
+                                message: "reading hex include",
+                                path: <PathBuf as Into<PathBuf>>::into(hex_path.clone()).to_owned(),
+                            }
                         })?;
 
                     let raw = hex::decode(file.trim())
                         .map_err(|e| Box::new(e) as Box<dyn std::error::Error>)
                         .context(error::InvalidHex {
-                            path: partial.path().to_owned(),
+                            path: <PathBuf as Into<PathBuf>>::into(hex_path).to_owned(),
                         })?;
 
-                    partial.push(vec![Node::Raw(raw)]);
+                    raws.push(RawOp::Raw(raw))
                 }
             }
         }
 
-        if !self.sources.sources.is_empty() {
-            panic!("extra sources?");
-        }
+        Ok(raws)
+    }
 
-        Ok(())
+    fn resolve_and_ingest<P: AsRef<Path>>(
+        &mut self,
+        program: &mut Program,
+        path: P,
+    ) -> Result<Vec<RawOp>, Error>
+    where
+        P: Into<PathBuf>,
+    {
+        let path = path.into();
+        let source = program.push_path(path.clone())?;
+        let code = read_to_string::<&PathBuf>(&source).with_context(|_| error::Io {
```

Comment:
Is the turbofish (`::<>`) necessary here?

Reply:
To be honest, the compiler asked me to add it so that the code can compile.  But I just removed it as you said and now there is no error. Possibly related to my first implementation. I'm sorry

Reply:
Don't be sorry, that's what reviews are for!